File: tools/kb-create/v2/catalog/graph.go
package catalog

import (
	"fmt"
	"sort"
	"strings"

	"github.com/kb-labs/create/v2/contracts"
)

// CompatibilityGraphSchema versions the release-owned compatibility graph.
// Version 3 replaces the flat label matrix of version 2: a cartesian
// SDK x platform x binary product cannot express service profiles, provider
// capabilities or per-target variants without growing a new column per
// dimension, while a graph absorbs a new node kind without a schema change.
const CompatibilityGraphSchema = "kb.release-compatibility/3"

// Node kinds. `binary` is the only kind with an OS/arch variant; every other
// kind is addressed by the package identity that ships it.
const (
	KindPlatform  = "platform"
	KindMember    = "member"
	KindSDK       = "sdk"
	KindPlugin    = "plugin"
	KindAdapter   = "adapter"
	KindBinary    = "binary"
	KindService   = "service"
	KindLauncher  = "launcher"
	KindToolchain = "toolchain"
)

// Edge kinds. `requires` and `provides` are the two the TypeScript sealer
// emits today; `compatibleWith` and `conflictsWith` are read by the resolver
// so a future release can express a non-directional relation without a
// launcher release.
const (
	EdgeRequires       = "requires"
	EdgeProvides       = "provides"
	EdgeCompatibleWith = "compatibleWith"
	EdgeConflictsWith  = "conflictsWith"
)

// GraphNode is one addressable release participant. `Variant` carries a
// non-target dimension (service profile, provider capability) and `Digest`
// pins the exact artifact bytes when the node is directly installable.
type GraphNode struct {
	ID      string `json:"id"`
	Kind    string `json:"kind"`
	Version string `json:"version"`
	OS      string `json:"os,omitempty"`
	Arch    string `json:"arch,omitempty"`
	Variant string `json:"variant,omitempty"`
	Digest  string `json:"digest,omitempty"`
}

// GraphEdge relates two nodes by their keys. `Range` is an optional semver
// constraint carried for auditability; edge presence, not range arithmetic, is
// what the strict resolver accepts.
type GraphEdge struct {
	From  string `json:"from"`
	To    string `json:"to"`
	Kind  string `json:"kind"`
	Range string `json:"range,omitempty"`
}

// GraphProfile names the member set and provider set of one service profile.
type GraphProfile struct {
	ID        string   `json:"id"`
	Members   []string `json:"members"`
	Providers []string `json:"providers"`
}

type CompatibilityGraph struct {
	Schema   string         `json:"schema"`
	Nodes    []GraphNode    `json:"nodes"`
	Edges    []GraphEdge    `json:"edges"`
	Profiles []GraphProfile `json:"profiles,omitempty"`
}

// NodeKey mirrors releaseGraphNodeKey in the TypeScript release contracts
// byte-for-byte. A bundle sealed by the plugin and a graph traversed here must
// address the same node with the same string, otherwise every edge dangles.
func NodeKey(node GraphNode) string {
	if node.Kind == KindBinary {
		return fmt.Sprintf("binary:%s@%s:%s/%s", node.ID, node.Version, node.OS, node.Arch)
	}
	return fmt.Sprintf("package:%s@%s", node.ID, node.Version)
}

// PackageKey and BinaryKey are the two call sites that build a key without
// already holding a node.
func PackageKey(id, version string) string { return "package:" + id + "@" + version }
func BinaryKey(id, version, os, arch string) string {
	return fmt.Sprintf("binary:%s@%s:%s/%s", id, version, os, arch)
}

func validateGraph(graph CompatibilityGraph, source Catalog) error {
	if graph.Schema != CompatibilityGraphSchema {
		return contracts.ReleaseError(
			contracts.CodeReleaseSchemaUnsupported,
			contracts.StageResolve,
			"release compatibility graph schema is not supported",
			"republish the release with the current release plugin; there is no legacy graph reader",
			fmt.Errorf("unsupported compatibility graph schema %q, expected %q", graph.Schema, CompatibilityGraphSchema),
		)
	}
	if len(graph.Nodes) == 0 {
		return contracts.ReleaseError(contracts.CodeReleaseIndexInvalid, contracts.StageResolve,
			"release compatibility graph declares no nodes",
			"reseal the release index with a complete compatibility graph", nil)
	}
	known := make(map[string]GraphNode, len(graph.Nodes))
	for _, node := range graph.Nodes {
		if strings.TrimSpace(node.ID) == "" || strings.TrimSpace(node.Kind) == "" || strings.TrimSpace(node.Version) == "" {
			return contracts.ReleaseError(contracts.CodeReleaseIndexInvalid, contracts.StageResolve,
				"release compatibility node is incomplete",
				"every graph node must declare id, kind and version", nil)
		}
		if node.Kind == KindBinary {
			if !contracts.SupportedTarget(node.OS, node.Arch) {
				return contracts.ReleaseError(contracts.CodeReleaseTargetUnsupported, contracts.StageResolve,
					"release compatibility graph declares a binary outside the supported matrix",
					"supported targets are "+strings.Join(contracts.SupportedTargets(), ", "),
					fmt.Errorf("binary %q targets %s/%s", node.ID, node.OS, node.Arch),
				).WithDetail("target", node.OS+"/"+node.Arch)
			}
		} else if node.OS != "" || node.Arch != "" {
			return contracts.ReleaseError(contracts.CodeReleaseIndexInvalid, contracts.StageResolve,
				"only binary graph nodes may declare an os/arch variant",
				"remove os/arch from non-binary nodes or reclassify the node as a binary", nil)
		}
		key := NodeKey(node)
		if _, exists := known[key]; exists {
			return contracts.ReleaseError(contracts.CodeReleaseIndexInvalid, contracts.StageResolve,
				"release compatibility graph declares a duplicate node",
				"seal exactly one node per release participant",
				fmt.Errorf("duplicate graph node %q", key))
		}
		known[key] = node
		if err := graphNodeArtifactExists(source, node); err != nil {
			return err
		}
	}
	for _, edge := range graph.Edges {
		switch edge.Kind {
		case EdgeRequires, EdgeProvides, EdgeCompatibleWith, EdgeConflictsWith:
		default:
			return contracts.ReleaseError(contracts.CodeReleaseIndexInvalid, contracts.StageResolve,
				"release compatibility graph declares an unknown edge kind",
				"use requires, provides, compatibleWith or conflictsWith",
				fmt.Errorf("unknown edge kind %q", edge.Kind))
		}
		for _, endpoint := range []string{edge.From, edge.To} {
			if _, exists := known[endpoint]; !exists {
				return contracts.ReleaseError(contracts.CodeReleaseGraphNodeUnknown, contracts.StageResolve,
					"release compatibility graph references an unknown node",
					"reseal the release index; edges may only address nodes it declares",
					fmt.Errorf("edge %s -> %s references absent node %q", edge.From, edge.To, endpoint),
				).WithDetail("node", endpoint)
			}
		}
	}
	for _, profile := range graph.Profiles {
		if strings.TrimSpace(profile.ID) == "" {
			return contracts.ReleaseError(contracts.CodeReleaseIndexInvalid, contracts.StageResolve,
				"release compatibility profile must declare an id", "reseal the release index", nil)
		}
		for _, member := range append(append([]string(nil), profile.Members...), profile.Providers...) {
			if _, exists := known[member]; !exists {
				return contracts.ReleaseError(contracts.CodeReleaseGraphNodeUnknown, contracts.StageResolve,
					"release compatibility profile references an unknown node",
					"reseal the release index; profiles may only address declared nodes",
					fmt.Errorf("profile %q references absent node %q", profile.ID, member),
				).WithDetail("node", member)
			}
		}
	}
	return nil
}

// graphNodeArtifactExists keeps the graph honest about the index it ships in:
// a node that no shipped artifact backs would let the resolver approve a
// combination the release cannot actually install.
func graphNodeArtifactExists(source Catalog, node GraphNode) error {
	present := false
	switch node.Kind {
	case KindPlatform:
		for _, value := range source.Platforms {
			if value.Package == node.ID && value.Version == node.Version {
				present = true
			}
		}
	case KindMember:
		for _, platform := range source.Platforms {
			for _, member := range platform.Members {
				if member.Package == node.ID && member.Version == node.Version {
					present = true
				}
			}
		}
	case KindSDK:
		present = componentPresent(source.SDKs, node)
	case KindPlugin:
		present = componentPresent(source.Plugins, node)
	case KindAdapter:
		for _, adapter := range source.Adapters {
			if adapter.Package == node.ID && adapter.Version == node.Version {
				present = true
			}
		}
	case KindBinary:
		for _, platform := range source.Platforms {
			for _, binary := range platform.Binaries {
				if binary.ID == node.ID && platform.Version == node.Version && binary.OS == node.OS && binary.Arch == node.Arch {
					present = true
				}
			}
		}
	default:
		// service, launcher, toolchain and any kind added later are declared
		// by the release rather than shipped as an index artifact. Generic
		// validation must not reject a kind this launcher predates.
		return nil
	}
	if present {
		return nil
	}
	return contracts.ReleaseError(contracts.CodeReleaseGraphNodeUnknown, contracts.StageResolve,
		"release compatibility graph declares a node the release index does not ship",
		"reseal the release index so every graph node has a shipped artifact",
		fmt.Errorf("graph node %q has no %s artifact in the index", NodeKey(node), node.Kind),
	).WithDetail("node", NodeKey(node))
}

func componentPresent(values []Component, node GraphNode) bool {
	for _, value := range values {
		if value.Package == node.ID && value.Version == node.Version {
			return true
		}
	}
	return false
}

// CheckCompatibility is the runtime decision boundary. Selection is accepted
// only when the sealed graph explicitly relates the chosen nodes: semver
// proximity is not a compatibility claim for a pre-stable platform.
func CheckCompatibility(source Catalog, platformVersion, sdkVersion, binaryID, targetOS, targetArch string) error {
	if source.Compatibility == nil {
		return contracts.ReleaseError(contracts.CodeReleaseIndexInvalid, contracts.StageResolve,
			"release index has no compatibility graph",
			"install from a release index sealed with a compatibility graph", nil)
	}
	graph := *source.Compatibility
	platform, ok := findPlatform(source.Platforms, platformVersion)
	if !ok {
		return contracts.ReleaseError(contracts.CodeReleaseGraphNodeUnknown, contracts.StageResolve,
			"selected platform is not present in the release index",
			"select a platform version the release index ships",
			fmt.Errorf("platform version %q is absent", platformVersion))
	}
	platformKey := PackageKey(platform.Package, platform.Version)
	if !hasNode(graph, platformKey) {
		return unknownNode(platformKey)
	}
	selected := []string{platformKey}
	sdkKey := ""
	if sdkVersion != "" {
		sdk, found := findComponentByVersion(source.SDKs, sdkVersion)
		if !found {
			return contracts.ReleaseError(contracts.CodeReleaseGraphNodeUnknown, contracts.StageResolve,
				"selected SDK is not present in the release index",
				"select an SDK version the release index ships",
				fmt.Errorf("sdk version %q is absent", sdkVersion))
		}
		sdkKey = PackageKey(sdk.Package, sdk.Version)
		if !hasNode(graph, sdkKey) {
			return unknownNode(sdkKey)
		}
		if !related(graph, platformKey, sdkKey) {
			return missingEdge(platformKey, sdkKey)
		}
		selected = append(selected, sdkKey)
	}
	if binaryID != "" {
		if !contracts.SupportedTarget(targetOS, targetArch) {
			return contracts.ReleaseError(contracts.CodeReleaseTargetUnsupported, contracts.StageResolve,
				"the local platform is outside the supported release matrix",
				"supported targets are "+strings.Join(contracts.SupportedTargets(), ", "),
				fmt.Errorf("target %s/%s is not released", targetOS, targetArch),
			).WithDetail("target", targetOS+"/"+targetArch)
		}
		binaryKey := BinaryKey(binaryID, platform.Version, targetOS, targetArch)
		if !hasNode(graph, binaryKey) {
			return unknownNode(binaryKey)
		}
		if !related(graph, binaryKey, platformKey) {
			return missingEdge(binaryKey, platformKey)
		}
		if sdkKey != "" && !related(graph, binaryKey, sdkKey) {
			return missingEdge(binaryKey, sdkKey)
		}
		selected = append(selected, binaryKey)
	}
	return checkConflicts(graph, selected)
}

func checkConflicts(graph CompatibilityGraph, selected []string) error {
	chosen := make(map[string]bool, len(selected))
	for _, key := range selected {
		chosen[key] = true
	}
	for _, edge := range graph.Edges {
		if edge.Kind == EdgeConflictsWith && chosen[edge.From] && chosen[edge.To] {
			return contracts.ReleaseError(contracts.CodeIncompatibleComponents, contracts.StageResolve,
				"the release graph declares the selected components as conflicting",
				"choose a different platform, SDK or binary combination",
				fmt.Errorf("%s conflicts with %s", edge.From, edge.To))
		}
	}
	return nil
}

func hasNode(graph CompatibilityGraph, key string) bool {
	for _, node := range graph.Nodes {
		if NodeKey(node) == key {
			return true
		}
	}
	return false
}

// related accepts a directed requires/provides edge in either direction and a
// symmetric compatibleWith edge. The sealer emits `requires` from the dependent
// to its dependency; the resolver only needs to know the relation was declared.
func related(graph CompatibilityGraph, from, to string) bool {
	for _, edge := range graph.Edges {
		if edge.Kind == EdgeConflictsWith {
			continue
		}
		if (edge.From == from && edge.To == to) || (edge.From == to && edge.To == from) {
			return true
		}
	}
	return false
}

func unknownNode(key string) error {
	return contracts.ReleaseError(contracts.CodeReleaseGraphNodeUnknown, contracts.StageResolve,
		"release compatibility graph has no node for the selected component",
		"select a component combination the release actually sealed",
		fmt.Errorf("graph node %q is absent", key)).WithDetail("node", key)
}

func missingEdge(from, to string) error {
	return contracts.ReleaseError(contracts.CodeReleaseGraphEdgeMissing, contracts.StageResolve,
		"release compatibility graph does not relate the selected components",
		"select a component combination the release declares as compatible",
		fmt.Errorf("no edge relates %q to %q", from, to)).WithDetail("from", from).WithDetail("to", to)
}

func findComponentByVersion(values []Component, version string) (Component, bool) {
	for _, value := range values {
		if value.Version == version {
			return value, true
		}
	}
	return Component{}, false
}

// GraphNodeKeys is a stable rendering of the graph for journals and dossiers.
func GraphNodeKeys(graph CompatibilityGraph) []string {
	result := make([]string, 0, len(graph.Nodes))
	for _, node := range graph.Nodes {
		result = append(result, NodeKey(node))
	}
	sort.Strings(result)
	return result
}

File: plugins/release/manager-cli/src/shared/bundle/graph.ts
/**
 * Compatibility-graph construction for `provenance.json`.
 *
 * The graph is what bundle verification rules 2 and 4 are checked against: every
 * shipped binary must have a node for its exact `{id, version, os, arch}`, every
 * edge must resolve, and every platform profile must be satisfiable from nodes
 * this bundle actually ships. Building it here — from the same classified
 * package list and binary set the index was built from — is what keeps the two
 * documents from drifting apart.
 *
 * SemVer ranges are produced with the shared implementation only (rule 7).
 */

import {
  releaseGraphNodeKey,
  type ReleaseCompatibilityGraph,
  type ReleasePackageClassification,
} from '@kb-labs/release-manager-contracts';
import semver from 'semver';

import type { NormalizedBinary } from './binary-manifest.js';

export interface GraphPackage {
  name: string;
  version: string;
  classification: ReleasePackageClassification;
}

export interface BuildCompatibilityGraphInput {
  packages: GraphPackage[];
  binaries: Array<Pick<NormalizedBinary, 'id' | 'os' | 'arch'> & { version: string }>;
  platformPackage: string;
  platformVersion: string;
  /** The SDK's declared peer range on the platform — the real compatibility edge. */
  sdkPeerRange: string;
  sdkPackage: string;
  sdkVersion: string;
  /** Packages that must be installable together with the platform. */
  memberPackages: string[];
}

/** Nodes exist for every package on a compatibility line; `deliveryOnly` has none by definition. */
const GRAPHED = new Set<ReleasePackageClassification>(['platform', 'member', 'sdk', 'plugin', 'adapter']);

function exactRange(version: string): string {
  const parsed = semver.parse(version);
  if (!parsed) { throw new Error(`cannot build a compatibility range from non-SemVer version ${version}`); }
  return parsed.format();
}

export function buildCompatibilityGraph(input: BuildCompatibilityGraphInput): ReleaseCompatibilityGraph {
  const graphed = input.packages
    .filter(pkg => GRAPHED.has(pkg.classification))
    .sort((left, right) => left.name.localeCompare(right.name));

  const nodes: ReleaseCompatibilityGraph['nodes'] = graphed.map(pkg => ({
    id: pkg.name,
    kind: 'package' as const,
    version: pkg.version,
  }));

  const binaries = [...input.binaries].sort((left, right) =>
    `${left.id}:${left.os}/${left.arch}`.localeCompare(`${right.id}:${right.os}/${right.arch}`));

  for (const binary of binaries) {
    nodes.push({ id: binary.id, kind: 'binary', version: binary.version, os: binary.os, arch: binary.arch });
  }

  const packageNode = (name: string, version: string): string =>
    releaseGraphNodeKey({ id: name, kind: 'package', version });
  const platformKey = packageNode(input.platformPackage, input.platformVersion);

  if (semver.validRange(input.sdkPeerRange) === null) {
    throw new Error(`sdk peer range is not a valid SemVer range: ${input.sdkPeerRange}`);
  }

  const edges: ReleaseCompatibilityGraph['edges'] = [];

  // The SDK's peer range is the one genuinely declared constraint between the
  // two version lines; everything else on the platform line ships lockstep, so
  // its edges pin the exact shipped version rather than inventing a range.
  edges.push({
    from: packageNode(input.sdkPackage, input.sdkVersion),
    to: platformKey,
    kind: 'requires',
    range: input.sdkPeerRange,
  });

  for (const pkg of graphed) {
    if (pkg.name === input.platformPackage || pkg.name === input.sdkPackage) { continue; }
    edges.push({
      from: packageNode(pkg.name, pkg.version),
      to: platformKey,
      kind: 'requires',
      range: exactRange(input.platformVersion),
    });
  }

  for (const binary of binaries) {
    edges.push({
      from: releaseGraphNodeKey({ id: binary.id, kind: 'binary', version: binary.version, os: binary.os, arch: binary.arch }),
      to: platformKey,
      kind: 'provides',
      range: exactRange(input.platformVersion),
    });
  }

  // One profile per shipped binary target: a profile is "what a consumer on this
  // os/arch installs", so its providers are that target's binaries and its
  // members are every package on the platform line plus the SDK.
  const memberKeys = graphed
    .filter(pkg => pkg.classification !== 'deliveryOnly')
    .filter(pkg => input.memberPackages.length === 0
      || pkg.name === input.platformPackage
      || pkg.name === input.sdkPackage
      || input.memberPackages.includes(pkg.name)
      || pkg.classification === 'plugin'
      || pkg.classification === 'adapter'
      || pkg.classification === 'member')
    .map(pkg => packageNode(pkg.name, pkg.version));

  const targets = [...new Set(binaries.map(binary => `${binary.os}-${binary.arch}`))].sort();
  const profiles: ReleaseCompatibilityGraph['profiles'] = targets.map(target => ({
    id: target,
    members: memberKeys,
    providers: binaries
      .filter(binary => `${binary.os}-${binary.arch}` === target)
      .map(binary => releaseGraphNodeKey({
        id: binary.id, kind: 'binary', version: binary.version, os: binary.os, arch: binary.arch,
      })),
  }));

  return { nodes, edges, profiles };
}
